vortex-layout/src/segments.rs

```diff
@@ -48,8 +56,223 @@ pub trait SegmentWriter {
     fn put(&mut self, buffer: &[ByteBuffer]) -> SegmentId;
 }
 
+#[derive(Debug, Default, PartialEq, Eq, Hash, Clone, Copy, PartialOrd, Ord)]
+pub enum RequiredSegmentKind {
+    PRUNING = 1,
+    FILTER = 2,
+    #[default]
+    PROJECTION = 3,
+}
+
+#[derive(Debug, PartialEq, Eq, Hash, Clone, Copy, PartialOrd, Ord)]
+pub struct SegmentPriority {
+    row_end: u64, // sort by row_end first
+    kind: RequiredSegmentKind,
+    row_start: u64,
+}
+
+impl SegmentPriority {
+    fn new(row_start: u64, row_end: u64, kind: RequiredSegmentKind) -> Self {
+        SegmentPriority {
+            row_end,
+            kind,
+            row_start,
+        }
+    }
+}
+
+const TOP_PRIORITY: SegmentPriority = SegmentPriority {
+    row_end: 0,
+    kind: RequiredSegmentKind::PRUNING,
+    row_start: 0,
+};
+
+type SegmentStore = BTreeMap<SegmentPriority, Vec<SegmentId>>;
+
+#[derive(Default)]
+pub struct SegmentCollector {
+    store: Arc<RwLock<SegmentStore>>,
+    pub kind: RequiredSegmentKind,
+}
+
+impl SegmentCollector {
+    pub fn with_priority_hint(&self, kind: RequiredSegmentKind) -> Self {
```

Comment:
This is a bit weird, that e.g. LayoutReaders are exposed to this API and have the possibility of calling it at all